src/middleware/immer.ts

```diff
@@ -0,0 +1,67 @@
+import { produce } from 'immer'
+import type { Draft } from 'immer'
+import { State, StateCreator, StoreMutatorIdentifier } from '../vanilla'
+
+type Immer = <
+  T extends State,
+  Mps extends [StoreMutatorIdentifier, unknown][] = [],
+  Mcs extends [StoreMutatorIdentifier, unknown][] = []
+>(
+  initializer: StateCreator<T, [...Mps, ['zustand/immer', never]], Mcs>
+) => StateCreator<T, Mps, [['zustand/immer', never], ...Mcs]>
+
+declare module 'zustand' {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  interface StoreMutators<S, A> {
+    ['zustand/immer']: WithImmer<S>
+  }
+}
+
+export type WithImmer<S> = S extends {
```

Comment:
Yes, we should deprecate `StoreWithDevtools` (and all other equivalent types that take state type instead of store type as parameter) and not export them in v4. I'll make that change in #772 too.